libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp

```diff
@@ -262,6 +262,20 @@ void NavEKF3_core::SelectMagFusion()
     // used for load levelling
     magFusePerformed = false;
 
+    // Handle case where we are using an external yaw sensor instead of a magnetomer
+    if ((frontend->_magCal == 5) && !use_compass()) {
+        if (storedYawAng.recall(yawAngDataDelayed,imuDataDelayed.time_ms)) {
+            if (tiltAlignComplete && !yawAlignComplete) {
+                alignYawAngle();
+            } else if (tiltAlignComplete && yawAlignComplete) {
+                fuseEulerYaw(false, true);
+            } else {
+                fuseEulerYaw(true, true);
+            }
+        }
```

Comment:
If we have no gps yaw and are on the ground,  shall we fuse in a synthetic heading measurement to prevent the
 filter covariances from becoming badly conditioned? Just like below:
    // If we have no magnetometer, fuse in a synthetic heading measurement at 7Hz to prevent the filter covariances
    // from becoming badly conditioned. For planes we only do this on-ground because they can align the yaw from GPS when
    // airborne. For other platform types we do this all the time.
    if (!use_compass()) {
        if ((onGround || !assume_zero_sideslip()) && (imuSampleTime_ms - lastSynthYawTime_ms > 140)) {
            fuseEulerYaw();
            magTestRatio.zero();
            yawTestRatio = 0.0f;
            lastSynthYawTime_ms = imuSampleTime_ms;
        }
    }

Reply:
When relying on GPS for yaw we should:

a) Not align the filter and start navigating until there is GPS.
b) If GPS yaw is lost, wait until a specified timeout period before fusing in synthetic yaw.

I will add patches for these changes.

Reply:
Can we fuse synthetic yaw before GPS yaw valid? Because most of GPS always take dozens of seconds to lock(fixed status) yaw,  just like NovAtel 617D.

Reply:
I've pushed a patch that amends the logic so that when using external yaw data it will align the tilt immediately and will start using synthetic yaw if no external yaw data has been received in the first 15 seconds.